code/Serving_API.py

```diff
@@ -1,27 +1,258 @@
+
+    return coordCityMap, coord_set, G
+
 
 app = FastAPI()
 
+
 @app.get("/physical-route/")
 def physical_route(src_latitude: float, src_longitude: float,
                    dst_latitude: float, dst_longitude: float) -> list[tuple[float, float]]:
-    """Get the physical route in (lat, lon) format from src to dst, including both ends.
-
-    Returns:
-        A list of (lat, lon) tuples, one for each hop along the route. This should include both ends.
     """
-    # TODO: get the real physical routes from database.
-    return [
-        (src_latitude, src_longitude),
-        (0, 0),
-        (dst_latitude, dst_longitude),
-    ]
+    Get the physical route in (lat, lon) format from src to dst, including both ends.
+    """
+    # Convert input coordinates to city information
+    src_city_co = find_closest_point(
+        (src_latitude, src_longitude), coord_set)
+
+    dst_city_co = find_closest_point(
+        (dst_latitude, dst_longitude), coord_set)
+
+    src_city_info = coordCityMap[src_city_co]
+    dst_city_info = coordCityMap[dst_city_co]
+
+    assert src_city_info and dst_city_info
+
+    # Find shortest path between cities in the graph
+    try:
+        shortest_path_cities = nx.shortest_path(
+            G, source=src_city_info, target=dst_city_info, weight='distance')
+
+        shortest_distance, coordinate_list, wkt_list, cable_type_list = calculate_shortest_path_distance(
+            G, shortest_path_cities)
+
+        return (coordinate_list, wkt_list, cable_type_list)
+    except nx.NetworkXNoPath:
+        print("cannot find the shortest path")
+        return ([], [], [])  # or handle the error as you prefer
+
 
 def run():
     import uvicorn
     uvicorn.run(app, port=8082)
 
+
+def calculate_shortest_path_distance(G, shortest_path_cities):
+    total_distance = 0
+    coordinate_list = []
+    wkt_list = []
+    cable_type_list = []
+
+    for i in range(len(shortest_path_cities) - 1):
+        city1 = shortest_path_cities[i]
+        city2 = shortest_path_cities[i + 1]
+
+        total_distance += G[city1][city2]['weight']
+        # Only append the coordinates of the source city here
+        coordinate_list.append(G.nodes[city1]['coord'])
+        wkt_list.append(G[city1][city2]['path_wkt'])
+        cable_type_list.append(G[city1][city2]['cable_type'])
+
+    # Append the coordinates of the last city after the loop
+    coordinate_list.append(G.nodes[shortest_path_cities[-1]]['coord'])
+
+    return total_distance, coordinate_list, wkt_list, cable_type_list
+
+
+def city_formatter(city_info):
+    city, state, country = city_info
+    return (city.strip(), state.strip(), country.strip())
+
+
+def parse_csv(filename):
+    data = []
+    with open(filename, newline='') as csvfile:
+        reader = csv.DictReader(csvfile)
+        for row in reader:
+            data.append(row)
+    return data
+
+
+def function_tester():
```

Comment:
Can you follow [this](https://fastapi.tiangolo.com/tutorial/testing/) to move testing code to a separate file?

This also avoids the duplicate code logic inside the API handler function `physical_route`.